src/System.Management.Automation/engine/Interop/Windows/GetDosDevice.cs

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#nullable enable
+
+#if !UNIX
+using System;
+using System.Buffers;
+using System.ComponentModel;
+using System.Management.Automation;
+using System.Runtime.InteropServices;
+
+internal static partial class Interop
+{
+    internal static partial class Windows
+    {
+        [LibraryImport(PinvokeDllNames.QueryDosDeviceDllName, EntryPoint = "QueryDosDeviceW", StringMarshalling = StringMarshalling.Utf16, SetLastError = true)]
+        internal static partial int QueryDosDevice(Span<char> lpDeviceName, Span<char> lpTargetPath, uint ucchMax);
+
+        internal static string GetDosDevice(char deviceName)
+        {
+            // By default buffer size is set to 300 which would generally be sufficient in most of the cases.
+            const int StartLength =
+#if DEBUG
+                // In debug, validate ArrayPool growth.
+                1;
+#else
+                300;
+#endif
+
+            Span<char> buffer = stackalloc char[StartLength + 1];
+            Span<char> fullDeviceName = stackalloc char[3] { deviceName, ':', '\0' };
+            char[]? rentedArray = null;
+
+            try
+            {
+                while (true)
+                {
+                    uint length = (uint)buffer.Length;
+                    int retValue = QueryDosDevice(fullDeviceName, buffer, length);
+                    if (retValue > 0)
+                    {
+                        if (buffer.StartsWith("\\??\\"))
+                        {
+                            // QueryDosDevice always return array of NULL-terinating strings with additional final NULL
+                            // so the buffer has always two NULL-s on end.
+                            //
+                            // "\\??\\UNC\\localhost\\c$\\tmp\0\0" -> "UNC\\localhost\\c$\\tmp\0\0"
+                            Span<char> res = buffer.Slice(4);
+                            if (res.StartsWith("UNC"))
+                            {
+                                // -> "C\\localhost\\c$\\tmp\0\0" -> "\\\\localhost\\c$\\tmp"
+                                res = res.Slice(2, retValue - 2);
+                                res[0] = '\\';
+
+                                // If we want always to have terminating slash -> "\\\\localhost\\c$\\tmp\\"
+                                // res = res.Slice(2, retValue - 3);
+                                // res[0] = '\\';
+                                // res[^1] = '\\';
+                            }
+                            else if (res[^3] == ':')
```

Comment:
Should we change this to an assert?  The comment is correct that this API is only ever used for network paths

Reply:
Added.

Reply:
If this method is to be used more widely in the future, it will have to be renamed again. If renaming, then it is logical to comment out the dead code and replace it with throw. Thoughts?

Reply:
Since it seems unlikely this will be used in the future, I would suggest rename and throw.

Reply:
Done.
(File renamed to pinvoke name as in other PSs.)